### ISSUE TITLE
npm task fails to authenticate (works when using command line task instead)

### ISSUE THREAD

**Issue by jdshkolnik** [microsoft/azure-pipelines-tasks#8628]
## Environment
- Server - Azure Pipelines or TFS on-premises? Both
    
    - If using TFS on-premises, provide the version: 16.131.28106.2
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: Deloitte-FinTech, Solvas

- Agent - Hosted or Private: Private

## Issue Description

We have a private Artifactory instance. The agent service account's profile has an .npmrc file with a valid authentication token. When I log in as that account and run npm install from command line, it works. When I create a command task in the build and run it, it works. When I use the npm task, it fails. When I explicitly tell it to use a service connection with valid authentication, it fails.

In short, something is wrong with the npm task's ability to authenticate. This isn't a failure of npm.

**Comment by jdshkolnik** [comment]
[npm-debug.txt](https://github.com/Microsoft/azure-pipelines-tasks/files/2485114/npm-debug.txt)

**Comment by jotaylo** [comment]
Just a quick guess before I start digging deeper, but your registries are specified with a port in the npm configuration, but the port isn't there in the service connection (https://private.artifactory:443/artifactory/api/npm/npm/ vs https://private.artifactory/artifactory/api/npm/npm/)
Can you please try updating the connection to include the port as well.

**Comment by alexmullans** [comment]
@jdshkolnik were you ablet to get this working with @jotaylo's suggestion?

**Comment by jdshkolnik** [comment]
Does it need to be there? I wouldn't think so. It'd be the same as specifying port 80 for all http URLs.

**Comment by alexmullans** [comment]
@jdshkolnik nope, it's not needed. the challenge here is in the discrepancy. EIther remove the port from the `.npmrc` or add it to the service connection URL. The issue here is that our tasks matches URLs between teh config file and the service connection to know which credential/token to put where. It's unable to determine a match here due to the mismatched URLs.

**Comment by chapterjason** [comment]
Facing the same issue.

In the example I replaced the URL with <URL>, the port with <PORT> and the scope with <OURSCOPE>.
The URLs are the same, but the npm Authenticate says that `No service connections that matched <URL>:<PORT> were found`.

.npmrc in the project:
```
registry=http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/
always-auth=true
```

Task: npm Authenticate ./.npmrc -> SUCCESS -> but I don't think so...
```
2019-04-05T09:15:00.3070101Z ##[section]Starting: npm Authenticate ./.npmrc
2019-04-05T09:15:00.3218053Z ==============================================================================
2019-04-05T09:15:00.3218053Z Task         : npm Authenticate (for task runners)
2019-04-05T09:15:00.3218053Z Description  : Don't use this task if you're also using the npm task. Provides npm credentials to an .npmrc file in your repository for the scope of the build. This enables npm task runners like Gulp and Grunt to authenticate with private registries.
2019-04-05T09:15:00.3218053Z Version      : 0.0.10
2019-04-05T09:15:00.3218053Z Author       : Microsoft Corporation
2019-04-05T09:15:00.3218053Z Help         : 
2019-04-05T09:15:00.3218053Z ==============================================================================
2019-04-05T09:15:01.1968057Z Adding authentication to the .npmrc file at E:\_work\239\s\.npmrc
2019-04-05T09:15:01.2124322Z No service connections that matched <URL>:<PORT> were found
2019-04-05T09:15:01.2124322Z ##[section]Finishing: npm Authenticate ./.npmrc
```

Task: npm install -> FAIL
```
2019-04-05T09:15:02.4936879Z ##[section]Starting: npm install
2019-04-05T09:15:02.4936879Z ==============================================================================
2019-04-05T09:15:02.4936879Z Task         : npm
2019-04-05T09:15:02.4936879Z Description  : Install and publish npm packages, or run an npm command. Supports npmjs.com and authenticated registries like Package Management.
2019-04-05T09:15:02.4936879Z Version      : 1.0.30
2019-04-05T09:15:02.4936879Z Author       : Microsoft Corporation
2019-04-05T09:15:02.4936879Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613746)
2019-04-05T09:15:02.4936879Z ==============================================================================
2019-04-05T09:15:03.1655685Z [command]C:\Windows\system32\cmd.exe /D /S /C "E:\NodeJS\npm.cmd --version"
2019-04-05T09:15:04.1812042Z 6.4.1
2019-04-05T09:15:06.4624644Z [command]C:\Windows\system32\cmd.exe /D /S /C "E:\NodeJS\npm.cmd config list"
2019-04-05T09:15:07.4937216Z ; cli configs
2019-04-05T09:15:07.4937216Z metrics-registry = "http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/"
2019-04-05T09:15:07.4937216Z scope = ""
2019-04-05T09:15:07.4937216Z user-agent = "npm/6.4.1 node/v10.15.0 win32 x64"
2019-04-05T09:15:07.4937216Z 
2019-04-05T09:15:07.4937216Z ; environment configs
2019-04-05T09:15:07.4937216Z userconfig = "E:\\_work\\239\\npm\\2535.npmrc"
2019-04-05T09:15:07.4937216Z 
2019-04-05T09:15:07.4937216Z ; project config E:\_work\239\s\.npmrc
2019-04-05T09:15:07.4937216Z always-auth = true
2019-04-05T09:15:07.4937216Z registry = "http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/"
2019-04-05T09:15:07.4937216Z 
2019-04-05T09:15:07.4937216Z ; globalconfig E:\NodeJS\etc\npmrc
2019-04-05T09:15:07.4937216Z @<OURSCOPE>-types:registry = "http://<URL>:9090"
2019-04-05T09:15:07.4937216Z @<OURSCOPE>:registry = "http://<URL>:9090"
2019-04-05T09:15:07.4937216Z cache = "E:\\NodeJS-Data\\NPM\\Cache"
2019-04-05T09:15:07.4937216Z prefix = "E:\\NodeJS-Data\\NPM\\Prefix"
2019-04-05T09:15:07.4937216Z 
2019-04-05T09:15:07.4937216Z ; node bin location = E:\NodeJS\node.exe
2019-04-05T09:15:07.4937216Z ; cwd = E:\_work\239\s
2019-04-05T09:15:07.4937216Z ; HOME = C:\Users\TFSAgent
2019-04-05T09:15:07.4937216Z ; "npm config ls -l" to show all defaults.
2019-04-05T09:15:07.4937216Z 
2019-04-05T09:15:07.4937216Z [command]C:\Windows\system32\cmd.exe /D /S /C "E:\NodeJS\npm.cmd install"
2019-04-05T09:15:09.4937329Z npm ERR! code E401
2019-04-05T09:15:09.4937329Z npm ERR! 401 Unauthorized: semver@^6.0.0
2019-04-05T09:15:09.4937329Z 
2019-04-05T09:15:09.4937329Z npm ERR! A complete log of this run can be found in:
2019-04-05T09:15:09.4937329Z npm ERR!     E:\NodeJS-Data\NPM\Cache\_logs\2019-04-05T09_15_09_462Z-debug.log
2019-04-05T09:15:09.5093589Z Found npm debug log, make sure the path matches with the one in npm's output: E:\NodeJS-Data\NPM\Cache\_logs\2019-04-05T09_15_09_462Z-debug.log
2019-04-05T09:15:09.5093589Z 0 info it worked if it ends with ok
2019-04-05T09:15:09.5093589Z 1 verbose cli [ 'E:\\NodeJS\\node.exe',
2019-04-05T09:15:09.5093589Z 1 verbose cli   'E:\\NodeJS\\node_modules\\npm\\bin\\npm-cli.js',
2019-04-05T09:15:09.5093589Z 1 verbose cli   'install' ]
2019-04-05T09:15:09.5093589Z 2 info using npm@6.4.1
2019-04-05T09:15:09.5093589Z 3 info using node@v10.15.0
2019-04-05T09:15:09.5093589Z 4 verbose npm-session 5c3ec50b2a40025b
2019-04-05T09:15:09.5093589Z 5 silly install runPreinstallTopLevelLifecycles
2019-04-05T09:15:09.5093589Z 6 silly preinstall @<OURSCOPE>/git-utils@0.1.2-DEV
2019-04-05T09:15:09.5093589Z 7 info lifecycle @<OURSCOPE>/git-utils@0.1.2-DEV~preinstall: @<OURSCOPE>/git-utils@0.1.2-DEV
2019-04-05T09:15:09.5093589Z 8 silly install loadCurrentTree
2019-04-05T09:15:09.5093589Z 9 silly install readLocalPackageData
2019-04-05T09:15:09.5093589Z 10 timing stage:loadCurrentTree Completed in 15ms
2019-04-05T09:15:09.5093589Z 11 silly install loadIdealTree
2019-04-05T09:15:09.5093589Z 12 silly install cloneCurrentTreeToIdealTree
2019-04-05T09:15:09.5093589Z 13 timing stage:loadIdealTree:cloneCurrentTree Completed in 0ms
2019-04-05T09:15:09.5093589Z 14 silly install loadShrinkwrap
2019-04-05T09:15:09.5093589Z 15 timing stage:loadIdealTree:loadShrinkwrap Completed in 0ms
2019-04-05T09:15:09.5093589Z 16 silly install loadAllDepsIntoIdealTree
2019-04-05T09:15:09.5093589Z 17 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@types%2fnode 78ms
2019-04-05T09:15:09.5093589Z 18 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@types%2fsemver 78ms
2019-04-05T09:15:09.5093589Z 19 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@types%2fshelljs 78ms
2019-04-05T09:15:09.5093589Z 20 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/typescript 62ms
2019-04-05T09:15:09.5093589Z 21 silly fetchPackageMetaData error for @types/node@^11.13.0 401 Unauthorized: @types/node@^11.13.0
2019-04-05T09:15:09.5093589Z 22 silly fetchPackageMetaData error for @types/semver@^6.0.0 401 Unauthorized: @types/semver@^6.0.0
2019-04-05T09:15:09.5093589Z 23 silly fetchPackageMetaData error for @types/shelljs@^0.8.4 401 Unauthorized: @types/shelljs@^0.8.4
2019-04-05T09:15:09.5093589Z 24 silly fetchPackageMetaData error for typescript@^3.4.1 401 Unauthorized: typescript@^3.4.1
2019-04-05T09:15:09.5093589Z 25 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/yargs 62ms
2019-04-05T09:15:09.5093589Z 26 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/semver 62ms
2019-04-05T09:15:09.5093589Z 27 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/shelljs 62ms
2019-04-05T09:15:09.5093589Z 28 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@types%2fyargs 78ms
2019-04-05T09:15:09.5093589Z 29 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@sourecode%2fdependency-injection 62ms
2019-04-05T09:15:09.5093589Z 30 silly fetchPackageMetaData error for yargs@^13.2.2 401 Unauthorized: yargs@^13.2.2
2019-04-05T09:15:09.5093589Z 31 silly fetchPackageMetaData error for semver@^6.0.0 401 Unauthorized: semver@^6.0.0
2019-04-05T09:15:09.5093589Z 32 silly fetchPackageMetaData error for shelljs@^0.8.3 401 Unauthorized: shelljs@^0.8.3
2019-04-05T09:15:09.5093589Z 33 silly fetchPackageMetaData error for @types/yargs@^12.0.11 401 Unauthorized: @types/yargs@^12.0.11
2019-04-05T09:15:09.5093589Z 34 silly fetchPackageMetaData error for @sourecode/dependency-injection@^0.1.1 401 Unauthorized: @sourecode/dependency-injection@^0.1.1
2019-04-05T09:15:09.5093589Z 35 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/semver 16ms
2019-04-05T09:15:09.5093589Z 36 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/shelljs 0ms
2019-04-05T09:15:09.5093589Z 37 silly fetchPackageMetaData error for semver@^6.0.0 401 Unauthorized: semver@^6.0.0
2019-04-05T09:15:09.5093589Z 38 silly fetchPackageMetaData error for shelljs@^0.8.3 401 Unauthorized: shelljs@^0.8.3
2019-04-05T09:15:09.5093589Z 39 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/typescript 0ms
2019-04-05T09:15:09.5093589Z 40 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/yargs 0ms
2019-04-05T09:15:09.5093589Z 41 http fetch GET 401 http://<URL>:<PORT>/tfs/SharePoint/_packaging/Examples/npm/registry/@sourecode%2fdependency-injection 16ms
2019-04-05T09:15:09.5093589Z 42 silly fetchPackageMetaData error for typescript@^3.4.1 401 Unauthorized: typescript@^3.4.1
2019-04-05T09:15:09.5093589Z 43 silly fetchPackageMetaData error for yargs@^13.2.2 401 Unauthorized: yargs@^13.2.2
2019-04-05T09:15:09.5093589Z 44 silly fetchPackageMetaData error for @sourecode/dependency-injection@^0.1.1 401 Unauthorized: @sourecode/dependency-injection@^0.1.1
2019-04-05T09:15:09.5093589Z 45 timing stage:rollbackFailedOptional Completed in 0ms
2019-04-05T09:15:09.5093589Z 46 timing stage:runTopLevelLifecycles Completed in 422ms
2019-04-05T09:15:09.5093589Z 47 silly saveTree @<OURSCOPE>/git-utils@0.1.2-DEV
2019-04-05T09:15:09.5093589Z 48 verbose stack Error: 401 Unauthorized: semver@^6.0.0
2019-04-05T09:15:09.5093589Z 48 verbose stack     at fetch.then.res (E:\NodeJS\node_modules\npm\node_modules\pacote\lib\fetchers\registry\fetch.js:42:19)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at tryCatcher (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Promise._settlePromiseFromHandler (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\promise.js:512:31)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Promise._settlePromise (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\promise.js:569:18)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Promise._settlePromise0 (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\promise.js:614:10)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Promise._settlePromises (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\promise.js:693:18)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Async._drainQueue (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\async.js:133:16)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Async._drainQueues (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\async.js:143:10)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at Immediate.Async.drainQueues [as _onImmediate] (E:\NodeJS\node_modules\npm\node_modules\bluebird\js\release\async.js:17:14)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at runCallback (timers.js:705:18)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at tryOnImmediate (timers.js:676:5)
2019-04-05T09:15:09.5093589Z 48 verbose stack     at processImmediate (timers.js:658:5)
2019-04-05T09:15:09.5093589Z 49 verbose cwd E:\_work\239\s
2019-04-05T09:15:09.5093589Z 50 verbose Windows_NT 6.3.9600
2019-04-05T09:15:09.5093589Z 51 verbose argv "E:\\NodeJS\\node.exe" "E:\\NodeJS\\node_modules\\npm\\bin\\npm-cli.js" "install"
2019-04-05T09:15:09.5093589Z 52 verbose node v10.15.0
2019-04-05T09:15:09.5093589Z 53 verbose npm  v6.4.1
2019-04-05T09:15:09.5093589Z 54 error code E401
2019-04-05T09:15:09.5093589Z 55 error 401 Unauthorized: semver@^6.0.0
2019-04-05T09:15:09.5093589Z 56 verbose exit [ 1, true ]
2019-04-05T09:15:09.5093589Z 
2019-04-05T09:15:09.5249835Z ##[error]Error: Npm failed with return code: 1
2019-04-05T09:15:09.5249835Z ##[section]Finishing: npm install
```

**Comment by chapterjason** [comment]
With the hostname only it works, with the FQDN it fails.

tfs-1:8080 -> works
tfs-1.core.company.com:8080 -> does not work

**Comment by elbatk** [comment]
Hi @chapterjason , is it the case where matches were and mismatches do not? Or if you put the FQDN into both it still fails?

**Comment by aasim** [comment]
@shubham90 , can you please work with @phil-hodgson to figure out if the issue is reproducible and provide a fix if it is.
